androidApp/src/main/java/com/mbta/tid/mbta_app/android/map/HomeMapView.kt

```diff
@@ -282,17 +292,108 @@ fun HomeMapView(
 
             val context = LocalContext.current
 
+            val locationProvider = remember {
+                object : LocationProvider {
+                    private val consumers = mutableSetOf<LocationConsumer>()
+
+                    fun sendLocation(location: Point) {
+                        for (consumer in consumers) {
+                            consumer.onLocationUpdated(location)
+                        }
+                    }
+
+                    override fun registerLocationConsumer(locationConsumer: LocationConsumer) {
+                        consumers.add(locationConsumer)
+                    }
+
+                    override fun unRegisterLocationConsumer(locationConsumer: LocationConsumer) {
+                        consumers.remove(locationConsumer)
+                    }
+                }
+            }
+
+            LaunchedEffect(locationDataManager) {
+                locationDataManager.currentLocation.collect { location ->
+                    if (location != null) {
+                        locationProvider.sendLocation(
+                            Point.fromLngLat(location.longitude, location.latitude)
+                        )
+                    }
+                }
+            }
+
             MapEffect(true) { map ->
-                mapViewportState.followPuck()
                 map.mapboxMap.addOnMapClickListener { point -> handleStopClick(map, point) }
+                map.location.setLocationProvider(locationProvider)
+            }
+
+            MapEffect(locationDataManager.hasPermission) { map ->
+                if (locationDataManager.hasPermission && viewportProvider.isDefault()) {
+                    viewportProvider.follow(
+                        DefaultViewportTransitionOptions.Builder().maxDurationMs(0).build()
+                    )
+                    layerManager.run { resetPuckPosition() }
+                }
+            }
+
+            LifecycleStartEffect(Unit) {
+                locationPermissions.launchMultiplePermissionRequest()
+                onStopOrDispose {}
+            }
+
+            LifecycleStartEffect(Unit) {
+                onStopOrDispose { viewportProvider.saveCurrentViewport() }
+            }
+
+            DisposableMapEffect { map ->
+                val listener =
+                    object : OnMoveListener, OnScaleListener, OnShoveListener {
```

Comment:
I also think this would benefit from being moved out of the file

Reply:
Moved.